api/turing/models/experiment_engine.go

```diff
@@ -6,25 +6,21 @@ import (
 	"errors"
 )
 
-// EngineType is used to capture the types of the supported experimentation engines
-// Need to expose this value in turing-router
-// engines/router/missionctl/experiment/experiment.go
-type ExperimentEngineType string
-
 const (
-	ExperimentEngineTypeNop    ExperimentEngineType = "nop"
-	ExperimentEngineTypeLitmus ExperimentEngineType = "litmus"
-	ExperimentEngineTypeXp     ExperimentEngineType = "xp"
+	ExperimentEngineTypeNop    string = "nop"
+	ExperimentEngineTypeLitmus string = "litmus"
+	ExperimentEngineTypeXp     string = "xp"
 )
 
 // ExperimentEngine contains the type and configuration for the
 // Experiment engine powering the router.
 type ExperimentEngine struct {
-	// Type of Experiment Engine. Currently supports "litmus", "xp" and "nop".
-	Type ExperimentEngineType `json:"type"`
+	// Type of Experiment Engine
+	Type string `json:"type"`
 	// Config contains the configs for the selected experiment engine (other than "nop").
-	// For standard experiment engine managers, the config can be cast into TuringExperimentConfig type.
-	Config interface{} `json:"config,omitempty"`
+	// For standard experiment engine managers, the config can be unmarshalled into
+	// manager.TuringExperimentConfig type.
+	Config json.RawMessage `json:"config,omitempty"`
```

Comment:
If using `json.RawMessage` here, the `Config` value saved in the DB will be a byte array, right?
Also, if I'm not wrong, in the current state of the code, the router / router version API would show the byte data.
I think it would be nice to see the actual config in both places.

Reply:
This would also require the existing data in our DB to be migrated (less of a concern). IMO `interface{}` in the GORM model is not a bad idea. I believe `map[string]interface{}` is more conventional for JSON.

Reply:
> If using json.RawMessage here, the Config value in the experiment_engine column in the DB will be displayed as a byte array, right?

yes, but it's not any different from how it works right now, because the entire  [ExperimentEngine](https://github.com/gojek/turing/blob/a1fa8c2d115f7a8fc28ac4b278bb7fa7d87bc0b4/api/turing/models/experiment_engine.go#L31) is stored as a byte array, not just `Config` value.

> Also, if I'm not wrong, in the current state of the code, the router / router version API would show the byte data.
I think it would be nice to see the actual config in both places.

This is something for me to check.

> This would also require the existing data in our DB to be migrated (less of a concern).

No, not required, because of the point above ^

> IMO interface{} in the GORM model is not a bad idea. I believe map[string]interface{} is more conventional for JSON.

So are you suggesting to use stick to `interface{}` here?

Reply:
> yes, but it's not any different from how it works right now

Hmmm.. It _is_ different. In the DB, the `experiment_engine` column is `jsonb` and we marshal the entire data before storing. However, when the jsonb column is queried, we see the decoded data.

But, if ExperimentEngine.Config is a byte array (`json.RawMessage`) in itself, I'm afraid we are going to see the same byte array in the DB, for example, through the `psql` command line.

CMIIW.

> So are you suggesting to use stick to interface{} here?

I now think `map[string]interface{}` is more apt (which is something I hadn't considered earlier, when changing the Config to `interface{}`) if we know it will be json data. Our OpenAPI specs state that the config field will be json.

Reply:
> Hmmm.. It _is_ different. In the DB, the `experiment_engine` column is `jsonb` and we marshal the entire data before storing. However, when the jsonb column is queried, we see the decoded data.
> 
> But, if ExperimentEngine.Config is a byte array (`json.RawMessage`) in itself, I'm afraid we are going to see the same byte array in the DB, for example, through the `psql` command line.
> 
> CMIIW.

I'm slightly confused by this. 
The type of `experiment_engine` remains `jsonb`, so you're expected to be able to query and see it as JSON data. Also, it doesn't matter if `ExperimentEngine.Config` is `interface{}` or `json.RawMessage` because in both cases the way how it is passed to the DB's driver is defined by the same [`ExperimentEngine.Value()`](https://github.com/gojek/turing/blob/a1fa8c2d115f7a8fc28ac4b278bb7fa7d87bc0b4/api/turing/models/experiment_engine.go#L30) method, which in both cases just marshals the entire `ExperimentEngine` instance (including its `Config` filed) into a byte array (`[]byte`). Essentially, the `Config` filed (no matter if it's `interface{}` or `json.RawMessage`) is being serialized into a byte array, which in case of `json.RawMessage` is even easier as it's already serialized.

> Also, if I'm not wrong, in the current state of the code, the router / router version API would show the byte data.
I think it would be nice to see the actual config in both places.

What I described above stands correct for the API logic as well. The API response [is serialized into a form of JSON](https://github.com/gojek/turing/blob/a1fa8c2d115f7a8fc28ac4b278bb7fa7d87bc0b4/api/turing/api/response.go#L32) using stdlib's json encoder, so the end result is the same for both `interface{}` and `json.RawMessage`.

Reply:
Querying `router_versions` of the existing deployment:
<img width="1552" alt="Screenshot 2021-12-21 at 7 48 54 PM" src="https://user-images.githubusercontent.com/1886194/147335043-40d252dc-46fb-4d01-9997-2282a7eb446d.png">

Querying `router_versions` of the local deployment with changes introduced in this PR:
<img width="1552" alt="Screenshot 2021-12-23 at 10 42 31 PM" src="https://user-images.githubusercontent.com/1886194/147335079-76e93e29-af2d-4e3f-a7ab-58593ac04795.png">

Reply:
After poking around a bit, I realise my source of confusion was in using `[]byte` vs `json.RawMessage`. The former shows something like:
<img width="1347" alt="Screenshot 2021-12-24 at 8 49 59 PM" src="https://user-images.githubusercontent.com/23465343/147354003-5289ffdc-4c6c-433f-9fa4-485af5b03b3a.png">

```
type ExperimentEngine struct {
	Type string `json:"type"`
	Config []byte `json:"config,omitempty"` // instead of json.RawMessage
}
```

This is something I had attempted earlier, before deciding to go with `interface{}` for the config. But, I had tried to double marshal the data, which apparently returns the base64 encoded string, and wrongly concluded that would be the same behavior for what happens when the data is saved to the DB. What I had failed to consider and take note of is, marshaling the `[]byte` behaves differently from marshaling `json.RawMessage`, even though the latter is an alias for the former.

A simplified example:
<img width="1009" alt="Screenshot 2021-12-24 at 9 01 08 PM" src="https://user-images.githubusercontent.com/23465343/147354568-7ec529d7-322e-42fd-af4e-6a5412eee7ee.png">

As you've elaborated, the `json.RawMessage` works as intended, both in the saved value in the DB as well the API response. Thanks for clarifying.